egs/librispeech/asr/simple_v1/mmi_bigram_embeddings_train.py

```diff
@@ -0,0 +1,746 @@
+    #  logging.info('the above crashes in the second batch (end)')
+    den_lats = k2.intersect_dense(den_graph, dense_fsa_vec, 10.0)
+
+    num_tot_scores = num_lats.get_tot_scores(
+        log_semiring=True,
+        use_double_scores=True)
+
+    den_tot_scores = den_lats.get_tot_scores(
+        log_semiring=True,
+        use_double_scores=True)
+
+    tot_scores = num_tot_scores - den_scale * den_tot_scores
+
+    (first_pass_tot_score, tot_frames,
+     all_frames) = get_tot_objf_and_num_frames(tot_scores,
+                                               supervision_segments[:, 2])
+
+    tot_score = first_pass_tot_score
+    if True:
+        # Now compute the tot_scores for the second pass
+        #
+        # TODO(fangjun): We probably need to split it into a separate function
+        padded_embeddings, len_per_path, path_to_seq, num_repeats = compute_embeddings(
+            den_lats,
+            graph_compiler.ctc_topo,
+            dense_fsa_vec,
+            max_phone_id=graph_compiler.max_phone_id,
+            num_paths=5, # NOTE(fangjun): a larger number results in OOM in `intersect_dense` below
+            debug=False)
+
+        # padded_embeddings is of shape [num_paths, max_phone_seq_len, num_features]
+        # i.e., [N, T, C]
+        padded_embeddings = padded_embeddings.permute(0, 2, 1)
+        # now padded_embeddings is [N, C, T]
+
+        #  logging.info(f'padded_embeddings {padded_embeddings.sum()}')
+
+        with torch.set_grad_enabled(is_training):
+            second_pass_out = second_pass_model(padded_embeddings)
+
+        assert second_pass_out.requires_grad == is_training
+
+        # second_pass_out is of shape [N, C, T]
+        second_pass_out = second_pass_out.permute(0, 2, 1)
+        # now second_pass_out is of shape [N, T, C]
+
+        #  logging.info(f'sum, {second_pass_out.sum()}')
+
+        assert second_pass_out.shape[0] == padded_embeddings.shape[0]
+        assert second_pass_out.shape[1] == padded_embeddings.shape[2]
+        assert second_pass_out.shape[2] == nnet_output.shape[2]
+
+        second_pass_supervision_segments = torch.stack(
+            (torch.arange(len_per_path.numel(), dtype=torch.int32),
+             torch.zeros_like(len_per_path), len_per_path),
+            dim=1)
+
+        indices = torch.argsort(len_per_path, descending=True)
+        assert indices.shape[0] == second_pass_supervision_segments.shape[0]
+        assert indices.shape[0] == path_to_seq.shape[0]
+
+        second_pass_supervision_segments = second_pass_supervision_segments[indices]
+        path_to_seq = path_to_seq[indices]
+        # no need to modify second_pass_out
+
+        num_repeats_float = k2.ragged.RaggedFloat(
+            num_repeats.shape(),
+            num_repeats.values().to(torch.float32))
+        path_weight = k2.ragged.normalize_scores(num_repeats_float,
+                                                 use_log=False).values
+        path_weight = path_weight[indices]
```

Comment:
this code is not super easy to follow...at some point I think some comments should be added, e.g. explaining the shapes of things and what they represent.

Reply:
Thanks, more comments have been added.